### PR TITLE
Fix artifact

### PR DESCRIPTION
**Pull request by hnpl**
- Fix the artifact object
- Decouple db and creating artifact

### CODE REVIEW

File: artifact/gem5art/artifact/artifact.py [hunk 1]
```diff
@@ -202,28 +197,56 @@ def registerArtifact(cls,
...
         data['url'] = url
-        data['supported_gem5_versions'] = supported_gem5_versions
+        data['supported_gem5_versions'] = supported_gem5_versions[:]
         data['version'] = version
 
-        for k, v in kwargs.items():
-            if k in data or k == "_id":
-                raise Exception("Field {} is reserved.".format(k))
-            data[k] = str(v)
+        data['extra'] = kwargs
```
Comment (powerjg) [inline artifact/gem5art/artifact/artifact.py:212]: Why are we no longer checking if it's reserved?

File: artifact/gem5art/artifact/artifact.py [hunk 2]
```diff
@@ -129,38 +129,33 @@ class Artifact:
...
+                       inputs: List['Artifact'] = [],
+                       architecture: str = "",
+                       size: int = None,
+                       is_zipped: bool = False,
+                       md5sum: str = "",
+                       url: str = "",
+                       supported_gem5_versions: List[str] = [],
+                       version: str = "",
+                       **kwargs: str
+                       ) -> 'Artifact':
```
Comment (powerjg) [inline artifact/gem5art/artifact/artifact.py:160]: Needs a docstring

File: artifact/gem5art/artifact/artifact.py [hunk 3]
```diff
@@ -202,28 +197,56 @@ def registerArtifact(cls,
...
+        _db = getDBConnection()
+
+        self = cls.createArtifact(command, name, cwd, typ, path, documentation,
+                                  inputs, architecture, size, is_zipped, md5sum,
+                                  url, supported_gem5_versions, version,
+                                  **kwargs)
 
-            # Now that we have a complete object, construct it
-            self = cls(data)
+        if self.hash in _db:
```
Comment (powerjg) [inline artifact/gem5art/artifact/artifact.py:252]: Should we do this before calling createArtifact?

File: artifact/gem5art/artifact/artifact.py [hunk 4]
```diff
@@ -260,6 +286,27 @@ def __init__(self, other: Union[str, UUID, Dict[str, Any]]) -> None:
         self.cwd = Path(other['cwd'])
         self.inputs = [Artifact(i) for i in other['inputs']]
 
+        # Optional fields
+        self.architecture = other['architecture'] if 'architecture' in other else ''
```
Comment (powerjg) [inline artifact/gem5art/artifact/artifact.py]: I think this whole thing can be simplified by doing the following
```suggestion
        self.architecture = other.get('architecture')
```

Note, you can use a second parameter to `get` to set a default. None probably works, though. See https://docs.python.org/3/library/stdtypes.html#dict.get

File: artifact/gem5art/artifact/artifact.py [hunk 5]
```diff
@@ -275,6 +322,8 @@ def _getSerializable(self) -> Dict[str, Union[str, UUID]]:
         data['inputs'] = [input._id for input in self.inputs]
         data['cwd'] = str(data['cwd'])
         data['path'] = str(data['path'])
+        data['supported_gem5_versions'] = str(self.supported_gem5_versions)
+        data['extra'] = str(self.extra)
```
Comment (powerjg) [inline artifact/gem5art/artifact/artifact.py]: Will `str` always work here? should we serialize extra? (I.e.,`json.dumps(extra)`)?

File: artifact/gem5art/artifact/artifact.py (hunk 1 above)
Comment (hnpl) [reply artifact/gem5art/artifact/artifact.py:212]: I put everything that is extra into 'extra', so having something like `extra.hash` wouldn't cause name collision issues.
However, do you think that we should enforce not having something like `extra.hash` anyway?

Comment (powerjg) [reply artifact/gem5art/artifact/artifact.py:212]: Gotcha. No, this makes sense. I'll resolve this.

File: artifact/gem5art/artifact/artifact.py (hunk 5 above)
Comment (hnpl) [reply artifact/gem5art/artifact/artifact.py]: fixed

**Review by powerjg — APPROVED** [review]
lgtm

**Pull request merged**